File: trunk/esper/src/test/java/net/esper/eql/join/assemble/TestSingleOptionalAssemblyNode.java
package net.esper.eql.join.assemble;

import net.esper.support.eql.join.SupportJoinProcNode;
import net.esper.support.eql.join.SupportJoinResultNodeFactory;
import net.esper.eql.join.rep.Node;
import net.esper.event.EventBean;

import java.util.List;

import junit.framework.TestCase;

public class TestSingleOptionalAssemblyNode extends TestCase
{
    private SupportJoinProcNode parentNode;
    private BranchOptionalAssemblyNode optAssemblyNode;
    private List<Node>[] resultMultipleEvents;
    private List<Node>[] resultSingleEvent;

    public void setUp()
    {
        optAssemblyNode = new BranchOptionalAssemblyNode(1, 4);
        parentNode = new SupportJoinProcNode(-1, 4);
        parentNode.addChild(optAssemblyNode);

        resultMultipleEvents = SupportJoinResultNodeFactory.makeOneStreamResult(4, 1, 2, 1); // 2 nodes 1 event each for (1)
        resultSingleEvent = SupportJoinResultNodeFactory.makeOneStreamResult(4, 1, 1, 1); // 1 nodes 1 event each for (1)
    }

    public void testProcessMultipleEvents()
    {
        optAssemblyNode.init(resultMultipleEvents);

        // generate an event row originating from a child for 1 of the 2 events in the result
        EventBean[] childRow = new EventBean[4];
        Node nodeOne = resultMultipleEvents[1].get(0);
        EventBean eventOne = nodeOne.getEvents().iterator().next();
        optAssemblyNode.result(childRow, 3, eventOne, nodeOne);

        // test that the node indeed manufactures event rows for any event not received from a child
        parentNode.getRowsList().clear();
        optAssemblyNode.process(resultMultipleEvents);

        // check generated row
        assertEquals(1, parentNode.getRowsList().size());
        EventBean[] row = parentNode.getRowsList().get(0);
        assertEquals(4, row.length);
        Node nodeTwo = resultMultipleEvents[1].get(1);
        assertEquals(nodeTwo.getEvents().iterator().next(), row[1]);
    }

    public void testProcessSingleEvent()
    {
        optAssemblyNode.init(resultSingleEvent);

        // test that the node indeed manufactures event rows for any event not received from a child
        optAssemblyNode.process(resultMultipleEvents);

        // check generated row
        assertEquals(1, parentNode.getRowsList().size());
        EventBean[] row = parentNode.getRowsList().get(0);
        assertEquals(4, row.length);
        Node node = resultSingleEvent[1].get(0);
        assertEquals(node.getEvents().iterator().next(), row[1]);
    }

    public void testChildResult()
    {
        optAssemblyNode.init(resultMultipleEvents);
        testChildResult(optAssemblyNode, parentNode);
    }

    protected static void testChildResult(BaseAssemblyNode nodeUnderTest, SupportJoinProcNode mockParentNode)
    {
        EventBean[] childRow = new EventBean[4];
        childRow[3] = SupportJoinResultNodeFactory.makeEvent();

        EventBean myEvent = SupportJoinResultNodeFactory.makeEvent();
        Node myNode = SupportJoinResultNodeFactory.makeNode(3, 1);

        // indicate child result
        nodeUnderTest.result(childRow, 3, myEvent, myNode);

        // assert parent node got the row
        assertEquals(1, mockParentNode.getRowsList().size());
        EventBean[] resultRow = mockParentNode.getRowsList().get(0);

        // assert the node has added his event to the row
        assertEquals(myEvent, resultRow[1]);
    }
}

File: NEsper.Tests/eql/join/assemble/TestSingleOptionalAssemblyNode.cs
using System;
using System.Collections.Generic;

using net.esper.eql.join.rep;
using net.esper.events;
using net.esper.support.eql.join;

using NUnit.Core;
using NUnit.Framework;

namespace net.esper.eql.join.assemble
{
	[TestFixture]
	public class TestSingleOptionalAssemblyNode
	{
		private SupportJoinProcNode parentNode;
		private BranchOptionalAssemblyNode optAssemblyNode;
		private IList<Node>[] resultMultipleEvents;
		private IList<Node>[] resultSingleEvent;

		[SetUp]
		public virtual void setUp()
		{
			optAssemblyNode = new BranchOptionalAssemblyNode( 1, 4 );
			parentNode = new SupportJoinProcNode( -1, 4 );
			parentNode.AddChild( optAssemblyNode );

			resultMultipleEvents = SupportJoinResultNodeFactory.makeOneStreamResult( 4, 1, 2, 1 ); // 2 nodes 1 event each for (1)
			resultSingleEvent = SupportJoinResultNodeFactory.makeOneStreamResult( 4, 1, 1, 1 ); // 1 nodes 1 event each for (1)
		}

		[Test]
		public virtual void testProcessMultipleEvents()
		{
			optAssemblyNode.Init( resultMultipleEvents );

			// generate an event row originating from a child for 1 of the 2 events in the result
			EventBean[] childRow = new EventBean[4];
			Node nodeOne = resultMultipleEvents[1][0];
			IEnumerator<EventBean> enumOne = nodeOne.Events.GetEnumerator();
			Assert.IsTrue( enumOne.MoveNext() );
			EventBean _eventOne = enumOne.Current;
			optAssemblyNode.Result( childRow, 3, _eventOne, nodeOne );

			// test that the node indeed manufactures event rows for any event not received from a child
			parentNode.getRowsList().Clear();
			optAssemblyNode.Process( resultMultipleEvents );

			// check generated row
			Assert.AreEqual( 1, parentNode.getRowsList().Count );
			EventBean[] row = parentNode.getRowsList()[0];
			Assert.AreEqual( 4, row.Length );
			Node nodeTwo = resultMultipleEvents[1][1];
			IEnumerator<EventBean> enumTwo = nodeTwo.Events.GetEnumerator();
			Assert.IsTrue( enumTwo.MoveNext() );
			Assert.AreEqual( enumTwo.Current, row[1] );
		}

		[Test]
		public virtual void testProcessSingleEvent()
		{
			optAssemblyNode.Init( resultSingleEvent );

			// test that the node indeed manufactures event rows for any event not received from a child
			optAssemblyNode.Process( resultMultipleEvents );

			// check generated row
			Assert.AreEqual( 1, parentNode.getRowsList().Count );
			EventBean[] row = parentNode.getRowsList()[0];
			Assert.AreEqual( 4, row.Length );
			Node node = resultSingleEvent[1][0];
			IEnumerator<EventBean> enumObj = node.Events.GetEnumerator();
			Assert.IsTrue( enumObj.MoveNext() );
			Assert.AreEqual( enumObj.Current, row[1] );
		}

		[Test]
		public virtual void testChildResult()
		{
			optAssemblyNode.Init( resultMultipleEvents );
			testChildResult( optAssemblyNode, parentNode );
		}

		protected internal static void testChildResult( BaseAssemblyNode nodeUnderTest, SupportJoinProcNode mockParentNode )
		{
			EventBean[] childRow = new EventBean[4];
			childRow[3] = SupportJoinResultNodeFactory.MakeEvent();

			EventBean myEvent = SupportJoinResultNodeFactory.MakeEvent();
			Node myNode = SupportJoinResultNodeFactory.MakeNode( 3, 1 );

			// indicate child result
			nodeUnderTest.Result( childRow, 3, myEvent, myNode );

			// assert parent node got the row
			Assert.AreEqual( 1, mockParentNode.getRowsList().Count );
			EventBean[] resultRow = mockParentNode.getRowsList()[0];

			// assert the node has added his event to the row
			Assert.AreEqual( myEvent, resultRow[1] );
		}
	}
}
